docs/spec/reactors/consensus/proposer-selection.md

```diff
@@ -2,20 +2,187 @@
-with the total voting power N, during a sequence of rounds of size N, every process is proposer in a number of rounds
-equal to its voting power.
+## Requirements for Proposer Selection
+
+The following requirements must be met by the Proposer Selection procedure:
+
+#### R1: Determinism
+Given a validator set `V`, and two honest validators `p` and `q`, for each height `h` and each round `r` the following must hold:
+
+  `proposer_p(h,r) = proposer_q(h,r)`
+
+where `proposer_p(h,r)` is the proposer returned by the Proposer Selection Procedure at process `p`, at height `h` and round `r`.
+
+#### R2: Liveness
+In every consecutive sequence of rounds of size K (K is system parameter), at least a
+single round has an honest proposer (??)
+
+#### R3: Fairness
+In a stable network (no validator set changes) and for any sequence of K rounds with K > P, a validator v is elected as proposer with a frequency f proportional to its voting power VP(v) divided by the total voting power P.
+f ~ VP(v) / P
+
+#### R4: Starvation Prevention ;)
+In a churning network with many validator set changes, minimal guarantees for chance in future election should be offered, e.g. for stable lower power validators.
+
+
+## Proposer Selection Specification
+__[R1]__ The proposer algorithm is deterministic giving consistent results accross executions with same transactions and validator set modifications. A pseudocode of the alorithm is given below. 
+
+__[R2]__ ??
+
+__[R3]__ Given a set of processes with the total voting power P, during a sequence of rounds of length P, every process is selected as proposer in a number of rounds equal to its voting power. The sequence of the P proposers then repeats.
+If we consider the validator set:
+
+Validator | p1 | p2 
+----------|--- | ---
+VP        | 1  | 3
+
+The proposer selection generates the sequence:
+`p2, p1, p2, p2, p2, p1, p2, p2,...` or [`p2, p1, p2, p2`]*
+
+The fairness of the proposer selection is ensured by assigning priorties to each validator based on the voting power and updating them at each round. In addition, everytime a validator is elected as proposer its priority is decreased with the total voting power.
+
+### Basic Algorithm
+
+This section includes an overview of the proposer selection algorithm. 
+
+A model that gives a good intuition on how/ why the selection algorithm works and it is fair is that of a priority queue. The validators move ahead in this queue according to their voting power (the higher the voting power the faster a validator moves towards the head of the queue). At each round the following happens:
+- all validators move "ahead" according to their powers: for each validator, increase the priority by the voting power
+- first in the queue becomes the proposer: select the validator with highest priority 
+- move the proposer to the end of the queue: decrease the proposer's priority by the total voting power
```

Comment:
Does it really always go to the end of the queue? For example, in cases we have consecutive proposals.

Reply:
Sometimes the same proposer proposes multiple times in a row, which means they're not necessarily pushed to the end of the queue after proposing, even if there's no validator changes.

Reply:
added the 'times' parameter and handling